### PR TITLE
Improve 'make' command substitution

### PR DESCRIPTION
**Pull request by alsgnet**
'make' may not be necessary the first word in the build command line. If it is then 'make' substitution fails and '-j' option is not passed to the 'make' command. For example MAKE directive array may be declared as:

MAKE[0]="source my_environment && make ..."

It causes parallel build to be ignored. The patch imroves parsing of the build command and substitutes 'make' regardless its position in the command line.

Change-Id: Id7aefdb5db87821849d5785cdc4b86cb69c259c4

### CODE REVIEW

**Comment by alsgnet** [comment]
I'm working on a distro that has vendor supplied kernels built with different compilers so I need a way to override the compiler toolchain. The driver build is called by invoking common.postinst script from a package. It is not manual build by calling the dkms script directly. The environment file (my_environment as it is in my example) is prepared by PRE_BUILD script by extracting compiler version string from a kernel that dkms is currently working on.

**Comment by alsgnet** [comment]
Regardless the fact that a suggested patch may resolve a problem with building a module on a system where installed kernels built with different compilers the patch actually does what is says in the title. It is more robust to parse the command line than relay on a sub-string matching. So, I would split the goals in two:

1. My patch as an improvement
2. Use compiler collection in dkms as it is indicated in #299 with an attempt to implement it in #298

Answering your questions, here is a simple demonstration of one of the solutions to switch compilers with PRE_BUILD script as a helper. Let's use an example in #299 on Oracle Linux 8 with RHEL gcc toolset. PRE_BUILD may contain something like this:

`
for f in $(find /opt/rh -type f -a -name gcc); do
    if strings /boot/vmlinuz-$KERNELVER | grep -q "$($f --version | head -1)"; then
        . ${f%/*}/../../../enable
         break
    fi
done
echo "PATH=$PATH" >my_environment
`
and "my_environment" can be used in conjunction with make:

`
MAKE[0]=". my_environment && make ...
`

There is a bunch of variables that RHEL gcc toolset sets up in "enable" file but only PATH is needed to build a kernel module successfully. A Makefile you requested isn't important. It could be a simple as:

`
obj-m += hello.o
`
I briefly looked through #298 and I think the approach is OK but implementation isn't complete. The author is mostly focused on compiler detection logic but not on actual outcome of that detection. In #299 he suggested to create either links to "gcc" of use a wrapper in /usr/local/bin to make it work. Maybe it is OK for an individual developer but it is unacceptable for production use. Let's say if I'm serving some customers and I supply them with a packaged driver. What would they say if I requested them to create links to every gcc in the system prior to install my package? They would find another supplier, I guess.

I think the idea behind #298 implementation should be like iteration of all available compilers in the $PATH and setting PATH locally in dkms for the "make" based on detection instead of passing CC and LD. This logic is missing. I tried the patch in #298 and it works with a wrapper in /usr/local/bin. Without wrapper (or links) it doesn't pick the right gcc even if both are in the $PATH. Only the first found is used for all kernel in the system which is incorrect.

**Comment by alsgnet** [comment]
I'm not sure why text formatting is broken. Let me try again:
```

for f in $(find /opt/rh -type f -a -name gcc); do
    if strings /boot/vmlinuz-$KERNELVER | grep -q "$($f --version | head -1)"; then
        . ${f%/*}/../../../enable
        break
    fi
done

echo "PATH=$PATH" >my_environment

```

**Comment by anbe42** [comment]
The documentation says
> KERNELRELEASE will be automatically appended to MAKE[#]. If you want to
> suppress this behavior, you can quote the make command: 'make'.
The same holds for injecting a -j# option, but is not documented.

Does this behavior still work with the proposed patch?

**Comment by anbe42** [comment]
Ideally we should set `MAKEFLAGS=-j$parallel_jobs` instead of "injecting" that flag into the MAKE command. At least if MAKEFLAGS is unset. And we probably shouldn't mess around with parallelism at all if there is an outer make already in control of it.

Which leaves the question: How can I write a dkms.conf which ensures that the MAKE command gets run "sequentially", i.e. with something equivalent to -j1 to build a module with a horribly broken build system? This needs to override any -j option given to dkms or autodetected $parallel_jobs value.

Not that I have a concrete example in mind that needs this, but we shouldn't force parallelism without having a way to opt out. Defaulting to parallelism is a good idea. (Manually running `dkms -j1 ...` is not a way to opt out with `dkms autoinstall` etc. being called from various hook scripts.)

**Comment by alsgnet** [comment]
@anbe42 ,

> Does this behavior still work with the proposed patch?

The patch doesn't change any current behavior. It only suggests to use more accurate parsing to drop a restriction that "make" has to be the first word in the command.

As for make options like "-j" or KERNELRELEASE=. In general it is not about options but a possibility to switch the compiler when building a dkms module in OS like Oracle 8 that uses different gcc for its installed by default kernels (4.18 and 5.15).

**Comment by TriMoon** [comment]
No matter if you guys decide if it is preferable to change the detection logic:
Please find a solution without adding `awk` as a dependency for the `dkms` package...

**Review by TriMoon — CHANGES_REQUESTED** [review]

File: dkms.in [hunk 1]
```diff
@@ -1076,14 +1076,22 @@ actual_build()
...
+        }'
 
     echo $""
     echo $"Building module:"
 
     invoke_command "$clean" "Cleaning build area" background
     echo $"DKMS make.log for $module-$module_version for kernel $kernelver ($arch)" >> "$build_log"
     date >> "$build_log"
-    local the_make_command="${make_command/#make/make -j$parallel_jobs KERNELRELEASE=$kernelver}"
+    local the_make_command=$(echo $make_command | awk -v jobs=$parallel_jobs -v ver=$kernelver "$make_subst")
```
[the remark this answers is not in the document]
Comment (alsgnet) [reply dkms.in:1094]: Why? To request such things you have to provide strong arguments that you didn't. Also, awk is already used by dkms script as well as many other tools such as grep, sed etc. Should dependencies be dropped on all of them?

[the remark this answers is not in the document]
Comment (TriMoon) [reply dkms.in:1094]: @alsgnet Stating what you claim and providing proof of that are two different things...
I don't see `awk` as a dependency for `dkms` on my system and nor like it to be one...

**Comment by alsgnet** [comment]
@evelikov , for now yes. Currently my make command in the MAKE[] array is like I described:

MAKE[0]=""source my_environment && make ..."

but I have to re-pass "-jxxx" and KERNELRELEASE=  to the "make" because substitution in dkms script doesn't work e.g:

MAKE[0]=""source my_environment && make -jxxx KERNELRELEASE= ..."

Generated by PRE_BUILD script "my_environment" file contains only PATH variable that could be either PATH=$PATH or PATH=/opt/rh/gcc-toolset-11/root/bin:$PATH depending on a kernel is being built.

File: dkms.in (hunk 1 above)
[the remark this answers is not in the document]
Comment (alsgnet) [reply dkms.in:1094]: @TriMoon , it is off-topic. Please raise your questions with a packages vendor which is Canonical in your case.

**Comment by anbe42** [comment]
What about
```
MAKE[0]="source my_environment && make ${parallel_jobs+-j$parallel_jobs} KERNELRELEASE=${kernelver} ..."
```
No need to reinvent the values, just use the variables provided by dkms in your custom command.

BTW, is there documentation which variables can be used in MAKE[0] etc?

**Comment by scaronni** [comment]
Parallel builds can now be adjusted by a variable in https://github.com/dell/dkms/pull/432
